File: src/app/api/routes/workflows.py
from fastapi import APIRouter, Depends, HTTPException

from src.app.contracts.requests import PlanRequest, ReviewRequest, TestPlanRequest
from src.app.contracts.responses import (
    PlanResponse,
    ReviewResponse,
    RunDetailResponse,
    TestPlanResponse,
    TraceResponse,
)
from src.app.core.auth import UserContext, get_user_context
from src.app.services.workflow_service import workflow_service

router = APIRouter()


@router.post("/plan", response_model=PlanResponse)
async def create_plan(request: PlanRequest, user: UserContext = Depends(get_user_context)) -> PlanResponse:
    return workflow_service.create_plan(request, user)


@router.post("/review", response_model=ReviewResponse)
async def create_review(request: ReviewRequest, user: UserContext = Depends(get_user_context)) -> ReviewResponse:
    return workflow_service.create_review(request, user)


@router.post("/test-plan", response_model=TestPlanResponse)
async def create_test_plan(request: TestPlanRequest, user: UserContext = Depends(get_user_context)) -> TestPlanResponse:
    return workflow_service.create_test_plan(request, user)


@router.get("/{run_id}", response_model=RunDetailResponse)
async def get_run(run_id: str, user: UserContext = Depends(get_user_context)) -> RunDetailResponse:
    try:
        result = workflow_service.get_run(run_id, user)
    except PermissionError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    if result is None:
        raise HTTPException(status_code=404, detail="run not found")
    return result


@router.get("/{run_id}/trace", response_model=TraceResponse)
async def get_trace(run_id: str, user: UserContext = Depends(get_user_context)) -> TraceResponse:
    try:
        result = workflow_service.get_trace(run_id, user)
    except PermissionError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    if result is None:
        raise HTTPException(status_code=404, detail="trace not found")
    return result

File: frontend/src/api/workflows.ts
import type { PlanRequest, ReviewRequest, TestPlanRequest } from "./contracts/requests";
import type { PlanResponse, ReviewResponse, RunDetailResponse, TestPlanResponse, TraceResponse } from "./contracts/responses";
import { fetchJson, type RequestContext } from "./http";

export function createPlan(payload: PlanRequest, context?: RequestContext) {
  return fetchJson<PlanResponse>("/v1/workflows/plan", { method: "POST", body: JSON.stringify(payload) }, context);
}

export function createReview(payload: ReviewRequest, context?: RequestContext) {
  return fetchJson<ReviewResponse>("/v1/workflows/review", { method: "POST", body: JSON.stringify(payload) }, context);
}

export function createTestPlan(payload: TestPlanRequest, context?: RequestContext) {
  return fetchJson<TestPlanResponse>("/v1/workflows/test-plan", { method: "POST", body: JSON.stringify(payload) }, context);
}

export function getRun(runId: string, context?: RequestContext) {
  return fetchJson<RunDetailResponse>(`/v1/workflows/${runId}`, undefined, context);
}

export function getTrace(runId: string, context?: RequestContext) {
  return fetchJson<TraceResponse>(`/v1/workflows/${runId}/trace`, undefined, context);
}
